### PR TITLE
Documentation for refactor with new Republisher job

### PR DESCRIPTION
**Pull request by jklukas**
The Republisher job proposed in this PR would factor out MarkAsSeen
from Decoder which would lead to a more logical flow of data.
It also allows us to share the expense of the MarkAsSeen read with
the read needed to inspect message contents and republish to
smaller topics.

Specifically, the Republisher for structured ingestion would check for
the new debug header and publish message containing that header to
a debug topic per the Glean request in #458.

The Republisher for telemetry data would randomly sample messages to
produce the monitoring topics discussed in #396.

### CODE REVIEW

File: docs/architecture/README.md [hunk 1]
```diff
@@ -81,13 +83,20 @@ This document specifies the architecture for GCP Ingestion as a whole.
       `geo_*` attributes
    1. Parse `agent` attribute into `user_agent_*` attributes
    1. Copy attributes to `metadata` top level key in `payload`
-   1. Remove duplicates based on `document_id` attribute using Cloud
-      Memorystore
+   1. Remove duplicates based on `document_id` attribute using `Cloud MemoryStore`
```
[the remark this answers is not in the document]
Comment (relud) [reply docs/architecture/README.md]: the reason this waits until the very last step to attempt deduplication is to minimize the likelyhood of duplicates due to the delay between checking an id and adding it to redis

File: docs/architecture/README.md [hunk 2]
```diff
@@ -81,13 +83,20 @@ This document specifies the architecture for GCP Ingestion as a whole.
...
       - Must ensure at least once delivery
-      - May read from output PubSub topics to determine delivered messages
 - Must send messages rejected by transforms to a configurable error destination
    - Must allow error destinations in PubSub and Cloud Storage
 
+### Republisher
+
+- Must copy messages from PubSub topics to PubSub topics
+- Must publish the `document_id` of each consumed message to `Cloud MemoryStore`
+- Must not ack messages read from PubSub until they are delivered
```
[the remark this answers is not in the document]
Comment (relud) [reply docs/architecture/README.md]:
```suggestion
- Must ack messages read from PubSub after they are delivered
- Must not ack messages read from PubSub before they are delivered
```

**Comment by jklukas** [comment]
> although this isn't changing much in this PR, marking messages in Redis is decoupled from publishing to `Decoded Topics`. This can lead to undetected duplicates in some cases (importance of this depends on how close in time are duplicates usually observed).

This actually isn't changing at all in this PR. The current state of the decoder is that it publishes to `Decoded Topics` and then reads from that topic to mark messages as read in Redis. The read will simply be happening in a different Dataflow job. I'm pushing a change now that includes some clarifying language about "best effort" deduplication and the time window between publishing a message and marking it as seen in Redis.

File: docs/architecture/README.md [hunk 3]
```diff
@@ -37,14 +37,16 @@ This document specifies the architecture for GCP Ingestion as a whole.
   `Cloud Storage`
 - The Dataflow `Decoder` job decodes messages from PubSub `Raw Topics` to
   PubSub `Decoded Topics`
-   - The Dataflow `Decoder` job uses `Cloud Memorystore` to deduplicate
-     messages
+   - The Dataflow `Decoder` job checks for existence of `document_id`s in
+     `Cloud Memorystore` in order to deduplicate messages
+- The Dataflow `Republisher` job reads messages from PubSub `Decoded Topics`,
+  marks them as seen in `Cloud Memorystore` and republishes to various
+  lower volume derived topics including `Monitoring Sample Topics` and
```
[the remark this answers is not in the document]
Comment (jklukas) [reply docs/architecture/README.md]: > These should be documented at some point.

The specific topic names, you mean?

File: docs/architecture/README.md [hunk 4]
```diff
@@ -81,13 +83,20 @@ This document specifies the architecture for GCP Ingestion as a whole.
...
 - Must send messages rejected by transforms to a configurable error destination
    - Must allow error destinations in PubSub and Cloud Storage
 
+### Republisher
+
+- Must copy messages from PubSub topics to PubSub topics
+- Must publish the `document_id` of each consumed message to `Cloud MemoryStore`
+- Must not ack messages read from PubSub until they are delivered
+   - Must ack messages that should not be delivered
+- Must accept configuration mapping `document_type`s to PubSub topics
```
[the remark this answers is not in the document]
Comment (jklukas) [reply docs/architecture/README.md]: Added detail here about the configuration parameters to accept for doc types and for sampling.

File: docs/architecture/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jklukas) [reply docs/architecture/README.md]: > the reason this waits until the very last step to attempt deduplication is to minimize the likelyhood of duplicates due to the delay between checking an id and adding it to redis

Added a line to document this concern.

File: docs/architecture/README.md [hunk 5]
```diff
@@ -81,13 +83,20 @@ This document specifies the architecture for GCP Ingestion as a whole.
...
+   1. Remove duplicates based on `document_id` attribute using `Cloud MemoryStore`
       - Must ensure at least once delivery
-      - May read from output PubSub topics to determine delivered messages
 - Must send messages rejected by transforms to a configurable error destination
    - Must allow error destinations in PubSub and Cloud Storage
 
+### Republisher
+
+- Must copy messages from PubSub topics to PubSub topics
+- Must publish the `document_id` of each consumed message to `Cloud MemoryStore`
```
[the remark this answers is not in the document]
Comment (jklukas) [reply docs/architecture/README.md]: Added some language about "best effort" for deduplication

File: docs/architecture/README.md [hunk 6]
```diff
@@ -81,13 +83,20 @@ This document specifies the architecture for GCP Ingestion as a whole.
...
+- Must not ack messages read from PubSub until they are delivered
+   - Must ack messages that should not be delivered
```
[the remark this answers is not in the document]
Comment (jklukas) [reply docs/architecture/README.md]: I've changed this language to reference "delivery to all matching destinations" and make no specific mention of "messages that should not be delivered"

File: docs/architecture/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (whd) [reply docs/architecture/README.md]: Yes, I'm filing a follow-up issue for formalizing our topic structure and naming conventions.

**Comment by jklukas** [comment]
Going ahead and merging at this point. Feel free to ask additional questions here, though. Opened #479 for implementing.

**Pull request merged**